A74-xds-config-tears.md

```diff
@@ -0,0 +1,311 @@
+    XdsEndpointResource endpoints;
+    std::string resolution_note;
+  };
+  std::map<std::string, absl::StatusOr<std::vector<ClusterConfig>>> clusters;
+};
+```
+
+Note that the watcher's OnError() and OnResourceDoesNotExist()
+methods will be invoked only for problems with the listener and route
+configuration resources.  For cluster-specific errors, we need to
+retain the existing behavior, which is that it causes a problem for the
+individual cluster but does not affect traffic sent to any other cluster.
+Therefore, when the XdsDependencyManager sees an error on a cluster for
+which it did not previously have data, or if it sees a does-not-exist
+notification on a cluster, it will report that by returning a non-OK
+status for the cluster in the XdsConfig.  Similarly, if there is a problem
+obtaining endpoint data from EDS or DNS, the XdsDependencyManager will
+report that problem via the cluster's resolution_note field.
+
+The xds resolver will pass the returned XdsConfig to the LB policies via
+an attribute, so that the LB policies can use the cluster data.  Note
+that we cannot simply pass the data via the cds LB policy's JSON config,
+because we need this to work in cases where the cluster is determined
+dynamically, such as where a route uses a ClusterSpecifierPlugin.  (We
+have some internal functionality that uses this, not something that has
+been documented in a public gRFC.)
+
+In order to handle such cases, the XdsDependencyManager will also provide
+an API to subscribe to a particular cluster resource whose name was
+not present in the route configuration.  The xds resolver will pass the
+XdsDependencyManager to the LB policies via an attribute, so that the
+cds LB policy can use this API.  The API will return some subscription
+handle that the caller can release when they are no longer interested
+in the cluster subscription.
+
+Note that in the XdsConfig, the cluster map will contain an entry for
+every cluster that is either referenced in the route configuration or
+for which a dynamic subscription was registered.  If a cluster is
+present only underneath an aggregate cluster, it will be present in the
+list for that aggregate cluster, but it will not have its own top-level
+entry.
+
+#### Changes in cds LB policy
+
+Instead of starting a watch on the XdsClient to get the cluster data,
+the cds LB policy will get its configuration by lookup up the cluster
+name in the XdsConfig passed to it via an attribute.
+
+To address the case where the cluster is determined dynamically (e.g., a
+route that uses a ClusterSpecifierPlugin), we will add a new field to
+the cds LB policy's configuration:
+
+```
+// Set this to true for a cluster that is determined dynamically
+// rather than being present in the route configuration.
+bool is_dynamic = 2;
+```
+
+If that field is set to true, then if the cluster name is not present in
+the XdsConfig, the cds policy will obtain a subscription handle from the
```

Comment:
The cds policy's config already has a field for the [cluster name](https://github.com/grpc/grpc-proto/blob/37330a01c4e1ac4d379bb81cc349df34aaf306cf/grpc/service_config/service_config.proto#L395).  When the ClusterSpecifierPlugin creates a cds policy for a given cluster, it sets that field so that the child policy knows which cluster it's for.

Today, that child cds policy gets the rest of the config by starting a watch on the XdsClient for that cluster name.  With this design, if the new `is_dynamic` field is set, then it will instead request a subscription handle for that policy from the XdsDependencyManager.  It will then wait for a subsequent address update (which the ClusterSpecifierPlugin will pass down to its children) that includes the updated xDS config with the new cluster.